pkg/jobs/controller.go

```diff
@@ -0,0 +1,350 @@
+	gen := Job{
+		dclient:       dclient,
+		queue:         workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), workQueueName),
+		dataStore:     newDataStore(),
+		configHandler: configHandler,
+		log:           log,
+	}
+	return &gen
+}
+
+func (j *Job) enqueue(info JobInfo) {
+	// add to data map
+	keyHash := info.toKey()
+	// add to
+	// queue the key hash
+
+	j.dataStore.add(keyHash, info)
+	j.queue.Add(keyHash)
+}
+
+//Add queues a policy violation create request
+func (j *Job) Add(infos ...JobInfo) {
+	for _, info := range infos {
+		j.enqueue(info)
+	}
+}
+
+// Run starts the workers
+func (j *Job) Run(workers int, stopCh <-chan struct{}) {
+	logger := j.log
+	defer utilruntime.HandleCrash()
+	logger.Info("start")
+	defer logger.Info("shutting down")
+	for i := 0; i < workers; i++ {
+		go wait.Until(j.runWorker, constant.PolicyViolationControllerResync, stopCh)
+	}
+	<-stopCh
+}
+
+func (j *Job) runWorker() {
+	for j.processNextWorkItem() {
+	}
+}
+
+func (j *Job) handleErr(err error, key interface{}) {
+	logger := j.log
+	if err == nil {
+		j.queue.Forget(key)
+		return
+	}
+
+	// retires requests if there is error
+	if j.queue.NumRequeues(key) < workQueueRetryLimit {
+		logger.Error(err, "failed to sync policy violation", "key", key)
+		// Re-enqueue the key rate limited. Based on the rate limiter on the
+		// queue and the re-enqueue history, the key will be processed later again.
+		j.queue.AddRateLimited(key)
+		return
+	}
+	j.queue.Forget(key)
+	// remove from data store
+	if keyHash, ok := key.(string); ok {
+		j.dataStore.delete(keyHash)
+	}
+	logger.Error(err, "dropping key out of the queue", "key", key)
+}
+
+func (j *Job) processNextWorkItem() bool {
+	logger := j.log
+	obj, shutdown := j.queue.Get()
+	if shutdown {
+		return false
+	}
+
+	err := func(obj interface{}) error {
+		defer j.queue.Done(obj)
+		var keyHash string
+		var ok bool
+
+		if keyHash, ok = obj.(string); !ok {
+			j.queue.Forget(obj)
+			logger.Info("incorrect type; expecting type 'string'", "obj", obj)
+			return nil
+		}
+
+		// lookup data store
+		info := j.dataStore.lookup(keyHash)
+		err := j.syncHandler(info)
+		j.handleErr(err, obj)
+		return nil
+	}(obj)
+
+	if err != nil {
+		logger.Error(err, "failed to process item")
+		return true
+	}
+
+	return true
+}
+
+func (j *Job) syncHandler(info JobInfo) error {
+	defer func() {
+		j.mux.Unlock()
+	}()
+	j.mux.Lock()
+
+	var wg sync.WaitGroup
+	if info.JobType == constant.BackgroundPolicySync {
+		wg.Add(1)
+		go j.syncKyverno(&wg, constant.All, constant.BackgroundPolicySync, info.JobData)
+	} else if info.JobType == constant.ConfiigmapMode {
+		if info.JobData != "" {
+			str := strings.Split(info.JobData, ",")
+			if len(str) > 1 {
+				wg.Add(1)
+				go j.syncKyverno(&wg, constant.All, constant.ConfiigmapMode, "")
+			} else {
+				wg.Add(len(str))
+				for _, scope := range str {
+					go j.syncKyverno(&wg, scope, constant.ConfiigmapMode, "")
+				}
+			}
```

Comment:
Is this `if len(str) > 1 {} else {}` correct? If `len() >1` we want to iterate through the slice.